app/models/shipit/webhooks/handlers/pull_request/review_stack_adapter.rb

```diff
@@ -29,6 +29,7 @@ def archive!(*args, &block)
             end
             return if stack.archived?
 
+            stack.remove_from_provisioning_queue if stack.awaiting_provision
```

Comment:
What happens if you don't include this conditional? Or if you move it inside `ReviewStack#remove_from_provisioning_queue`?

Reply:
> What happens if you don't include this conditional? 

Functionally, the conditional saves a few database calls, and Jobs when the ReviewStack is was not previously `awaiting_provisioning`:
 
```ruby
﻿2.7.1 :002 > review_stack.awaiting_provision
 => false
2.7.1 :003 > review_stack.remove_from_provisioning_queue
   (0.6ms)  BEGIN
  Shipit::Repository Exists? (0.9ms)  SELECT 1 AS one FROM `repositories` WHERE `repositories`.`name` = 'hello-world' AND `repositories`.`id` != 1 AND `repositories`.`owner` = 'codertocat' LIMIT 1
  Shipit::Stack Exists? (0.5ms)  SELECT 1 AS one FROM `stacks` WHERE `stacks`.`repository_id` = 1 AND `stacks`.`id` != 1 AND `stacks`.`environment` = 'pr2' LIMIT 1
   (0.4ms)  COMMIT
  Shipit::User Load (0.4ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1
  Shipit::PullRequest Load (0.7ms)  SELECT `pull_requests`.* FROM `pull_requests` WHERE `pull_requests`.`stack_id` = 1 LIMIT 1
  Shipit::Stack Load (0.9ms)  SELECT `stacks`.* FROM `stacks` WHERE `stacks`.`id` = 1 LIMIT 1
  Shipit::Repository Load (0.5ms)  SELECT `repositories`.* FROM `repositories` WHERE `repositories`.`id` = 1
  Shipit::User Load (0.6ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1
  Shipit::Commit Load (0.6ms)  SELECT `commits`.* FROM `commits` WHERE `commits`.`id` = 1 LIMIT 1
  Shipit::User Load (0.8ms)  SELECT `users`.* FROM `users` INNER JOIN `pull_request_assignments` ON `users`.`id` = `pull_request_assignments`.`user_id` WHERE `pull_request_assignments`.`pull_request_id` = 1
[ActiveJob] Enqueued Shipit::EmitEventJob (Job ID: 9cdef12d-7a2a-4a8a-abcf-63546dfb0126) to Sidekiq(hooks) with arguments: {:event=>"merge_status", :stack_id=>1, :payload=>"{\"merge_status\":\"locked\",\"stack\":{\"id\":1,\"repo_owner\":\"codertocat\",\"repo_name\":\"hello-world\",\"environment\":\"pr2\",\"html_url\":\"http://localhost:8080/codertocat/hello-world/pr2\",\"url\":\"http://localhost:8080/api/stacks/codertocat/hello-world/pr2\",\"tasks_url\":\"http://localhost:8080/api/stacks/codertocat/hello-world/pr2/tasks\",\"deploy_url\":null,\"merge_requests_url\":\"http://localhost:8080/api/stacks/codertocat/hello-world/pr2/merge_requests\",\"deploy_spec\":{\"merge\":{\"require\":[],\"ignore\":[],\"revalidate_after\":null,\"method\":null,\"max_divergence\":{\"commits\":null,\"age\":null}},\"ci\":{\"hide\":[],\"allow_failures\":[],\"require\":[],\"blocking\":[]},\"machine\":{\"environment\":{\"BUNDLE_PATH\":\"/home/app/src/data/bundler\"},\"directory\":null,\"cleanup\":true},\"review\":{\"checklist\":[],\"monitoring\":[],\"checks\":[]},\"plugins\":[\"failure-reasons\"],\"status\":{\"context\":null,\"delay\":\"0s\"},\"dependencies\":{\"override\":[]},\"provision\":{\"handler_name\":null},\"deploy\":{\"override\":null,\"variables\":[],\"max_commits\":8,\"interval\":\"0s\"},\"rollback\":{\"override\":null},\"fetch\":null,\"tasks\":{}},\"undeployed_commits_count\":1,\"is_locked\":true,\"lock_reason\":\"Archived\",\"continuous_deployment\":false,\"created_at\":\"2020-10-05T23:15:32.000Z\",\"updated_at\":\"2020-10-05T23:15:43.000Z\",\"locked_since\":\"2020-10-05T23:15:43.000Z\",\"last_deployed_at\":null,\"branch\":\"changes\",\"merge_queue_enabled\":false,\"is_archived\":true,\"archived_since\":\"2020-10-05T23:15:43.000Z\",\"lock_author\":{\"id\":1,\"name\":\"Codertocat\",\"email\":null,\"login\":\"Codertocat\",\"avatar_url\":\"https://avatars1.githubusercontent.com/u/21031067?v=4\",\"created_at\":\"2020-10-05T23:15:32.000Z\",\"updated_at\":\"2020-10-05T23:15:32.000Z\",\"github_id\":21031067,\"anonymous\":false},\"pull_request\":{\"id\":1,\"number\":2,\"title\":\"Update the README with new information.\",\"github_id\":279147437,\"additions\":1,\"deletions\":1,\"state\":\"open\",\"html_url\":\"https://github.com/codertocat/hello-world/pull/2\",\"user\":{\"id\":1,\"name\":\"Codertocat\",\"email\":null,\"login\":\"Codertocat\",\"avatar_url\":\"https://avatars1.githubusercontent.com/u/21031067?v=4\",\"created_at\":\"2020-10-05T23:15:32.000Z\",\"updated_at\":\"2020-10-05T23:15:32.000Z\",\"github_id\":21031067,\"anonymous\":false},\"head\":{\"sha\":\"ec26c3e57ca3a959ca5aad62de7213c562f8c821\",\"message\":\"Update README.md\"},\"assignees\":[]}}}"}
[Redis] command=SADD args="shipit:deferred_touches" ["Shipit::Repository|1|updated_at"]
[Redis] call_time=1.77 ms
[ActiveJob] Enqueued Shipit::DeferredTouchJob (Job ID: fbbbeff9-09ca-4c9a-b6b7-3637e6134e20) to Sidekiq(default)
 => true
2.7.1 :004 > review_stack.awaiting_provision
 => false
```

> Or if you move it inside ReviewStack#remove_from_provisioning_queue?

I hadn't considered moving the condition into the `ReviewStack#remove_from_provisioning_queue` method. At the moment, `remove_from_provisioning_queue` seems like a much more appropriate place to be doing the check. Going to give that a try. Thanks for the suggestion!

Reply:
Addressed by: https://github.com/Shopify/shipit-engine/pull/1106/commits/3e9356bc9a73da56b49f906fbf838a89dab93d38